src/js/lib/skeleton.js

```diff
@@ -71,38 +71,39 @@ class Skeleton extends EventEmitter {
     * @param {Boolean|String} noIpc - Flag to skip ipc emission or to do `both`.
     * @param {Boolean|String} [tabId=false] - Emit to specific tab over ipc.
     * @param {Boolean|String} [parent=false] - Emit to script's parent over ipc.
+    * @returns {Promise} - Promise that will resolve when the emit is done - if availabl
     */
     emit(event, data = {}, noIpc = false, tabId = false, parent = false) {
+        let promise
         if (this.env.isExtension && (!noIpc || noIpc === 'both')) {
             let payloadData = {data: data, event: event}
 
             if (tabId) {
                 if (this.verbose) this.logger.debug(`${this}emit ipc event '${event}' to tab ${tabId}`)
-                browser.tabs.sendMessage(tabId, payloadData).catch((err) => {
+                return browser.tabs.sendMessage(tabId, payloadData).catch((err) => {
                     if (this.verbose) this.logger.debug(`${this}${err.message}`)
                 })
-                return
             } else if (parent) {
                 if (this.verbose) this.logger.debug(`${this}emit ipc event '${event}' to parent`)
                 parent.postMessage({data: data, event: event}, '*')
-                return
+                return undefined
             }
 
             if (data && data.callback) {
                 if (this.verbose) this.logger.debug(`${this}emit ipc event with callback '${event}'`)
                 const callback = data.callback
                 // Make sure that functions are not part of the payload data.
                 delete data.callback
-                browser.runtime.sendMessage(payloadData).then(function handleResponse(message) {
+                promise = browser.runtime.sendMessage(payloadData).then(function handleResponse(message) {
                     callback(message)
                 }).catch((err) => {
                     if (this.verbose) this.logger.debug(`${this}${err.message}`)
                 })
             } else {
                 if (this.verbose) this.logger.debug(`${this}emit ipc event '${event}'`)
-                let _promise = browser.runtime.sendMessage(payloadData)
-                if (_promise) {
-                    _promise.catch((err) => {
+                promise = browser.runtime.sendMessage(payloadData)
+                if (promise) {
```

Comment:
It looks here as if browser.runtime.sendMessage doesn't always return a Promise, which it does.